feat: support read sn and read s33

little-willy opened the pull request:
In our current code of file manager, we have two method, `readS32` and `readS64`, which both read a variable length integer from the file. Actually, they seems to share the same read logic as described in [sN](https://webassembly.github.io/spec/core/binary/values.html#binary-int). Besides, as proposed in #2029, we may support reading `s33`, which also share the same read logic. 

Therefore, in this PR, I support reading `sN` as described in [sN](https://webassembly.github.io/spec/core/binary/values.html#binary-int), where N becomes a static constant template variable, and enables static usage check to ensure correctness. `readS32`, `readS64` and `readS33` are simply specialization of `readSN`.

juntao commented:
Hi @little-willy 

Thank you for your contribution!

CNCF requires its projects to have DCO signatures for all contributions. Do you mind fixing the DCO on your commits? Thank you!

https://www.secondstate.io/articles/dco/

q82419 commented on lib/loader/filemgr.cpp:
```diff
@@ -275,8 +287,8 @@ Expect<std::string> FileMgr::readName() {
-  // If UTF-8 validation or readU32() or readBytes() failed, the last succeeded
-  // reading offset will be at the start of `Name`.
+  // If UTF-8 validation or readU32() or readBytes() failed, the last
+  // succeeded reading offset will be at the start of `Name`.
```
This line is not necessary to be changed.

q82419 commented on lib/loader/filemgr.cpp:
```diff
@@ -411,7 +423,8 @@ Expect<void> FileMgr::jumpContent() {
 
-// Helper function for reading number of bytes. See "include/loader/filemgr.h".
+// Helper function for reading number of bytes. See
+// "include/loader/filemgr.h".
```
This line is not necessary to be changed.

q82419 reviewed (changes requested):
Can you help to fix the loading of block type [here](https://github.com/WasmEdge/WasmEdge/blob/master/lib/loader/ast/instruction.cpp#L188)?

Thanks!

q82419 commented on lib/loader/filemgr.cpp:
```diff
@@ -144,84 +144,96 @@ Expect<uint64_t> FileMgr::readU64() {
-    }
-    if (auto Res = testRead(1); unlikely(!Res)) {
+    WasmEdge::Byte Byte;
+    if (auto Res = readByte(); unlikely(!Res)) {
```
Please use the `testRead(1)` and `Byte = Data[Pos++]` mechanism as the origin code.
This is for printing the correct error message of the loading offset when error occurs.

little-willy commented:
@q82419 Thanks for reviewing. I have fixed the comment and added unit tests for `readS33`. PTAL.

> Can you help to fix the loading of block type [here](https://github.com/WasmEdge/WasmEdge/blob/master/lib/loader/ast/instruction.cpp#L188)?

Can I include this fix in a later PR that will fix #2029 totally and merge this PR first?

q82419 commented:
> @q82419 Thanks for reviewing. I have fixed the comment and added unit tests for `readS33`. PTAL.
> 
> > Can you help to fix the loading of block type [here](https://github.com/WasmEdge/WasmEdge/blob/master/lib/loader/ast/instruction.cpp#L188)?
> 
> Can I include this fix in a later PR that will fix #2029 totally and merge this PR first?

Sure.
Thanks for your contribution!
